packages/dds/map/src/mapKernel.ts

```diff
@@ -440,18 +444,43 @@ export class MapKernel {
         return false;
     }
 
+    /**
+     * Rollback a local op
+     * @param op - The operation to rollback
+     * @param localOpMetadata - The local metadata associated with the op.
+     */
+    public rollback(op: any, localOpMetadata: unknown) {
+        if (op.type !== "delete" && op.type !== "set") {
+            throw new Error("Unsupported op for rollback");
+        }
+
+        const metadata = localOpMetadata as IMapLocalOpMetadata;
+        if (metadata === undefined) {
+            this.deleteCore(op.key, true);
+        } else {
+            this.setCore(op.key, metadata.previousValue, true, true);
+        }
+
+        // BUGBUG: This isn't right. It should revert it to the value it was before the rolled back op.
+        this.pendingKeys.delete(op.key);
```

Comment:
I think the correct implementation would be for pendingKeys to keep a list of pending local op metadata per key. (I considered putting the previous value in the local op metadata, but that would make it hard for cleaning them up when there's a clear or ack.)

It would be nice to be able to know if we're within an orderSequentially with rollback to know to do all this extra work, but that's an optimization and maybe better not to have two different behaviours.

Reply:
i wouldn't worry about this optimization right now. we assume things will be fast, so the list will remain small.